stdlib/public/runtime/EnvironmentVariables.cpp

```diff
@@ -0,0 +1,165 @@
+  case '1':
+    return true;
+  case 'N':
+  case 'n':
+  case 'F':
+  case 'f':
+  case '0':
+    return false;
+  default:
+    swift::warning(RuntimeErrorFlagNone,
+                  "Warning: cannot parse value %s=%s, defaulting to %s.\n",
+                  name, value, defaultValue ? "true" : "false");
+    return defaultValue;
+  }
+}
+
+static uint8_t parse_uint8_t(const char *name,
+                             const char *value,
+                             uint8_t defaultValue) {
+  if (!value)
+    return defaultValue;
+  long n = strtol(value, nullptr, 0);
+  if (n == 0 && errno == EINVAL) {
+    swift::warning(RuntimeErrorFlagNone,
+                   "Warning: cannot parse value %s=%s, defaulting to %u.\n",
+                   name, value, defaultValue);
+    return defaultValue;
+  }
+
+  if (n < 0) {
+    swift::warning(RuntimeErrorFlagNone,
+                   "Warning: %s=%s out of bounds, clamping to 0.\n",
+                   name, value);
+    return 0;
+  }
+  if (n > UINT8_MAX) {
+    swift::warning(RuntimeErrorFlagNone,
+                   "Warning: %s=%s out of bounds, clamping to %d.\n",
+                   name, value, UINT8_MAX);
+    return UINT8_MAX;
+  }
+
+  return n;
+}
+
+// Print a list of all the environment variables. Lazy initialization makes
+// this a bit odd, but the use of these variables in the metadata system means
+// it's almost certain to run early.
+//
+// The "extra" parameter is printed after the header and before the list of
+// variables.
+static void printHelp(const char *extra) {
+  swift::warning(RuntimeErrorFlagNone, "Swift runtime debugging:\n");
+  if (extra)
+    swift::warning(RuntimeErrorFlagNone, "%s\n", extra);
+#define VARIABLE(name, type, defaultValue, help) \
+  swift::warning(RuntimeErrorFlagNone, "%7s %s [default: %s] - %s\n", \
+                 #type, #name, #defaultValue, help);
+#include "EnvironmentVariables.def"
+}
+
+// Initialization code.
+OnceToken_t Environment::initializeToken;
+
+#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__linux__)
+extern "C" char **environ;
+void Environment::initialize(void *context) {
+  // On platforms where we have an environment variable array available, scan it
+  // directly. This optimizes for the common case where no variables are set,
+  // since we only need to perform one scan to set all variables. It also allows
+  // us to detect some spelling mistakes by warning on unknown SWIFT_ variables.
+
+  bool SWIFT_HELP_variable = false;
+  for (char **var = environ; *var; var++) {
+    // Immediately skip anything without a SWIFT_ prefix.
+    if (strncmp(*var, "SWIFT_", 6) != 0)
+      continue;
+
+    bool foundVariable = false;
+#define VARIABLE(name, type, defaultValue, help)                       \
+    if (strncmp(*var, #name "=", strlen(#name "=")) == 0) {            \
+      name ## _variable =                                              \
+        parse_ ## type(#name, *var + strlen(#name "="), defaultValue); \
```

Comment:
A comment on how we are relying on the compile failure to indicate the unsupported type might be nice.  We could also do some trickery with the preprocessor and `static_assert` to identify an unsupported type.